商品出品機能

Wataru-Ki opened the pull request:
商品出品機能の実装

Wataru-Ki committed:
itemモデルの作成

Wataru-Ki committed:
ActiveStrage導入とそれに伴う調整

Wataru-Ki committed:
ActiveHashの導入と記述

Wataru-Ki committed:
itemモデル単体テストコード(金額の項を除く)

Wataru-Ki committed:
JavaScriptファイルの編集

Wataru-Ki committed:
itemモデル単体テストコード(追記)

s-aka-moto commented on app/models/item.rb:
```diff
@@ -0,0 +1,28 @@
+  has_one_attached :image
+
+  extend ActiveHash::Associations::ActiveRecordExtensions
+  belongs_to_active_hash :shipping_area
+  validates :shipping_area, presence: true
+  validates :shipping_area_id, numericality: { other_than: 1 }
+  belongs_to_active_hash :shipping_date
+  validates :shipping_date, presence: true
+  validates :shipping_date_id, numericality: { other_than: 1 }
+  belongs_to_active_hash :shipping_charge
+  validates :shipping_charge, presence: true
+  validates :shipping_charge_id, numericality: { other_than: 1 }
+  belongs_to_active_hash :category
+  validates :category, presence: true
+  validates :category_id, numericality: { other_than: 1 }
+  belongs_to_active_hash :item_condition
+  validates :item_condition, presence: true
+  validates :item_condition_id, numericality: { other_than: 1 }
+
+  validates :price, presence: true, numericality: { only_integer: true, greater_than: 300, less_than: 9999999 }
```
バリデーションは「with_options」を使用することで、共通したオプションを付けることが可能となります。

s-aka-moto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,74 @@
+
+  describe "商品出品機能" do
+    context "商品出品がうまくいくとき" do
+      it "入力必須項目を全て記入すれば登録できる" do
+        expect(@item).to be_valid
+      end
+    end
+  
+    context "商品出品がうまくいかないとき" do
+      it "商品名がないと登録できない" do
+        @item.name = ""
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Name can't be blank")
+      end
+      it "商品の説明がないと登録できない" do
+        @item.introduction = ""
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Introduction can't be blank")
+      end
+      it "カテゴリーの情報がないと登録できない" do
```
文字列として認識されてしまうため、クオテーションは外しましょう！
@item.category_id = `"1"`

s-aka-moto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,74 @@
+        expect(@item.errors.full_messages).to include("Price can't be blank")
+      end
+      it "価格が¥300以下だと登録できない" do
+        @item.price = "299"
```
こちらもクオテーションを外しましょう！
 @item.price =` "299"`

s-aka-moto commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,74 @@
+        expect(@item.errors.full_messages).to include("Price must be greater than 300")
+      end
+      it "価格が¥10,000,000以上だと登録できない" do
+        @item.price = "10000000"
```
こちらもクオテーションを外しましょう！
 @item.price = `"10000000"`

s-aka-moto reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

Wataru-Ki committed:
各ファイルの修正

①ダウンロードしたビューファイルのエラーメッセージを表示させるコードを正常に利用できるようにした。②バリデーションの記述をwith_optionsでまとめた。③Specファイル内の指定箇所のクォテーションマークの削除。

Wataru-Ki commented:
レビューありがとうございます！
修正いたしましたので再度確認お願いします！

i-date commented on app/javascript/item_price.js:
```diff
@@ -0,0 +1,19 @@
+function input (){
+//金額を入力した数値をpriceInputという変数に格納する
+const priceInput = document.getElementById("item-price");
+// console.log(priceInput);
```
4行目のデバッグ用のconsole.logは削除しましょう。

i-date commented on app/javascript/item_price.js:
```diff
@@ -0,0 +1,19 @@
+// console.log(priceInput);
+priceInput.addEventListener("input", () => {
+  const inputValue = priceInput.value;
+  // console.log(inputValue);
```
7行目のデバッグ用のconsole.logは削除しましょう。

i-date commented on app/javascript/item_price.js:
```diff
@@ -0,0 +1,19 @@
+
+  const addTaxDom = document.getElementById("add-tax-price");
+  addTaxDom.innerHTML = Math.floor(inputValue * 0.1)
+  // console.log(addTaxDom)
```
11行目のデバッグ用のconsole.logは削除しましょう。

i-date commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
-    <%# render 'shared/error_messages', model: f.object %>
+    <%= render 'shared/error_messages', model: f.object %>
     <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
10、12行目の下記の実装案内コメントですが、機能を実装いただいたため、こちらは不要となりますので削除しましょう。
不要になったコメントを削除することで、可読性や保守性を向上させることができます。
```
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
<%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
メモ代わりに残していただいている場合は修正は不要です。

i-date reviewed (commented):
修正ありがとうございます！
少し追加コメントいたしましたので、ご確認をお願いいたします。

i-date commented:
また、修正後の機能動作や画面を確認させていただきたいので、お手数ですがプルリクにキャプチャ動画（のURL）を載せていただき、再度プルリクエストをご提出いただけないでしょうか？
その際、「Gyazo GIF 使い方」とgoogleで検索し、GyazoGIFなどを使用して動画のURLを生成いただければと思います。

Wataru-Ki commented:
レビューありがとうございます！失礼いたしました、下記にGyazoのURLを記載いたします。
## 機能の様子
https://gyazo.com/b8125c31b05c3673bbbc490ee2e37bd5
## テストコードhttps://gyazo.com/a83980804572796f14ef8924abb88e67

Wataru-Ki committed:
修正

isyt729 commented:
レビュー依頼ありがとうございます！
LGTMといたします！
引き続き実装頑張りましょう！

Wataru-Ki commented:
ありがとうございます！
